### PR TITLE
Disable screen layout by default

### LINKED ISSUE #11716: Set screen layout mode off by default
### Is your feature request related to a problem? Please describe.
For an in-depth discussion, see: #11187

Essentially, screen layout can cause important information to be missed by users, the need for an understanding of screen layout is more specialized.

### Describe the solution you'd like
Make screen layout off by default in new config.

### PR DESCRIPTION
**Pull request by OzancanKaratas**
### Link to issue number:
Fixes #11716
### Summary of the issue:
A new NVDA user may not work with web pages.
### Description of how this pull request fixes the issue:
This pull request will disable the screen layout, making it easier to use the arrow keys on web pages.
### Testing performed:
Start NVDA in default settings and open a web page.
### Known issues with pull request:
Problems may occur with some Office programs that use IAccessible2. Please test Word while testing this pull request.
### Change log entry:

- Changes
    - NVDA will no longer use the screen layout feature by default.

### CODE REVIEW

**Comment by codeofdusk** [comment]
I vote to close this PR: disabling screen layout changes the way pages are presented. NVDA is a screen reader and shouldn't alter the presentation of the screen unless explicitly requested to do so. This would represent a significant departure in the way NVDA works.

**Comment by Brian1Gaff** [comment]
If you want to do this, surely the best plan is to make the choice one can 
make after testing the two modes. I've been happy with the current defaults 
we have.
 Maybe explain it in a quick guide more simply?
 Brian

bglists@blueyonder.co.uk
Sent via blueyonder.
Please address personal E-mail to:-
briang1@blueyonder.co.uk, putting 'Brian Gaff'
in the display name field.
Newsgroup monitored: alt.comp.blind-users
----- Original Message -----

**Comment by OzancanKaratas** [comment]
Yes @codeofdusk, you are right. However, A new NVDA user must first learn to use NVDA in order to change their previous habits. Changing the default setting here does not negatively affect NVDA and users. @leonardder, what do you think?

**Comment by lukaszgo1** [comment]
> Changing the default setting here does not negatively affect NVDA and users.

As demonstrated by the failing system test this change has a negative effect.
For the following HTML snipped:
`<p>Big <span aria-invalid="spelling">caat</span> meos</p>`
with screen layout disabled NVDA splits a line before the `span` starts but does not after it ends. My personal view is that even with screen layout disabled it should not split either before or after the `span`, as the text is entirely in one paragraph without any interactive controls which should be presented separately.
In other words adapting the failing system test to the new behavior is not the right way to address this build failure IMO.

**Comment by leonardder** [comment]
Note that screen layout of is a virtual buffer specific thing. It doesn't work in browse mode in Word, Excel, Outlook, Kindle, etc. Therefore this pr would introduce a difference in behaviour throughout the product depending on a factor that's hard to understand by users.

**Comment by Brian1Gaff** [comment]
If this is an issue it needs a separate number methinks, as whichever  way 
you make default that sounds like a bug. Brian

bglists@blueyonder.co.uk
Sent via blueyonder.
Please address personal E-mail to:-
briang1@blueyonder.co.uk, putting 'Brian Gaff'
in the display name field.
Newsgroup monitored: alt.comp.blind-users
----- Original Message -----

**Comment by OzancanKaratas** [comment]
@leonardder, can you help me if there is any other change that needs to be made?

**Comment by leonardder** [comment]
@OzancanKaratas Note from the several replies that there isn't overall consensus that this pr is a good idea. I leave it up to NV Access to decide that.

**Comment by amirsol81** [comment]
Honestly it's a great idea to keep Screen Layout disabled by default. When enabled, many pages are displayed so strangely that interaction with HTML elements becomes difficult. Note that I'm a long-time screen reader user.

**Comment by JulienCochuyt** [comment]
FWIW, here at Accessolutions, we disable screen layout for quite some time now on every new installation for customers, primarily because it allows for easier activation of different controls within the same paragraph.

cc @frederic-brugnot
cc @rthxy

**Comment by OzancanKaratas** [comment]
@JulienCochuyt, can you help me fix the test error?

**Comment by JulienCochuyt** [comment]
The failed tests are  "ARIA invalid spelling and grammar" and "Robot.chromeTests :: HTML test cases in Chrome".
They probably expect screen layout and need to be impacted.

**Comment by feerrenrut** [comment]
I think we should change the default for new users or configurations, but I'm a little unsure if it should be changed for existing users. The current implementation will affect anyone who hasn't manually changed the setting. I don't think this is quite what we want.

This will need a configSpec update step which should set the value explicitly to enabled for configs that are upgraded to the latest spec version. The idea here is that new users have new configs and thus won't have this update step run and will therefore just get the default value. Users who had the previous default, but had not explicitly configured it, will continue to get the behavior they are used to.

As for the system tests, I'll acknowledge that getting the results can be a little difficult with NVDA. Generally: you can get the error [by going to the test results tab on the appveyor build](https://ci.appveyor.com/project/NVAccess/nvda/builds/37472110/tests) and following the link for the broken test. Identifying the broken test is more difficult because of how they are exposed by Appveyor (visually they have a red mark next to them). However, failed tests are also links, so jumping to link should find them. Activating the link will expand a section showing the error message for the test. In this case it is:
```
Actual speech != Expected speech: Big != Big  spelling error  caat  meos
```

Another way to get the test results, including a full report and the NVDA logs for the test is to download the systemtest zip file from the artifacts tab. 

I suspect this test will need to be modified to press down arrow several more times and expect the speech piecemeal. 

Note that the system tests can be run locally.

**Comment by beqabeqa473** [comment]
i also vote to close this pr.

there is no benefit from this change.
user can do it manually if one will need it.

On 2/5/21, Reef Turner <notifications@github.com> wrote:
> I think we should change the default for new users or configurations, but
> I'm a little unsure if it should be changed for existing users. The current
> implementation will affect anyone who hasn't manually changed the setting. I
> don't think this is quite what we want.
>
> This will need a configSpec update step which should set the value
> explicitly to enabled for configs that are upgraded to the latest spec
> version. The idea here is that new users have new configs and thus won't
> have this update step run and will therefore just get the default value.
> Users who had the previous default, but had not explicitly configured it,
> will continue to get the behavior they are used to.
>
> As for the system tests, I'll acknowledge that getting the results can be a
> little difficult with NVDA. Generally: you can get the error [by going to
> the test results tab on the appveyor
> build](https://ci.appveyor.com/project/NVAccess/nvda/builds/37472110/tests)
> and following the link for the broken test. Identifying the broken test is
> more difficult because of how they are exposed by Appveyor (visually they
> have a red mark next to them). However, failed tests are also links, so
> jumping to link should find them. Activating the link will expand a section
> showing the error message for the test. In this case it is:
> ```
> Actual speech != Expected speech: Big != Big  spelling error  caat  meos
> ```
>
> Another way to get the test results, including a full report and the NVDA
> logs for the test is to download the systemtest zip file from the artifacts
> tab.
>
> I suspect this test will need to be modified to press down arrow several
> more times and expect the speech piecemeal.
>
> Note that the system tests can be run locally.
>
> --
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/nvaccess/nvda/pull/11906#issuecomment-773885298


-- 
with best regards Beqa Gozalishvili
Tell: +995593454005
Email: beqaprogger@gmail.com
Web: https://gozaltech.org
Skype: beqabeqa473
Telegram: https://t.me/gozaltech
facebook: https://facebook.com/gozaltech
twitter: https://twitter.com/beqabeqa473
Instagram: https://instagram.com/beqa.gozalishvili

**Comment by OzancanKaratas** [comment]
Older users familiar with the screen layout can re-enable the screen layout even if this setting is turned off. However, this pull request is aimed at new users who are completely unaware of this.

**Comment by feerrenrut** [comment]
NV Access is keen to see this go ahead. @OzancanKaratas please let us know if you need assistance.

**Comment by leonardder** [comment]
@feerrenrut could you please share your thoughts about my concerns in https://github.com/nvaccess/nvda/pull/11906#issuecomment-739800931

**Comment by CyrilleB79** [comment]
> Note that screen layout of is a virtual buffer specific thing. It doesn't work in browse mode in Word, Excel, Outlook, Kindle, etc. Therefore this pr would introduce a difference in behaviour throughout the product depending on a factor that's hard to understand by users.

I just want to indicate that Jaws does not use screen layout by default; I do not even know if Jaws supports a screen layout mode for virtual buffers. And I did not hear Jaws users complain about the difference between focus mode and browse mode.
IMO, for a beginner, the most important is generally to understand the structure of a document (e.g. there are two links) rather than knowing that visually the two links are on the same line.

**Comment by OzancanKaratas** [comment]
@feerrenrut, I disabled the screen layout via configSpec. However, help is needed to fix the test error.

**Comment by lukaszgo1** [comment]
@feerrenrut  I've created #12067 to discuss presentation of `aria-innvalid` with screen layout disabled as that will resolve the system test failure. Could you please take a look?
@CyrilleB79 In JAWS there is an option called "Document Presentation Mode" which can be set to screen layout which presents virtual buffers very similar to what NVDA does by default now.

**Comment by CyrilleB79** [comment]
> @CyrilleB79 In JAWS there is an option called "Document Presentation Mode" which can be set to screen layout which presents virtual buffers very similar to what NVDA does by default now.

OK thanks. From what I have read now, the "Simple layout" mode of Jaws, equivalent to "Screen mode off" of NVDA, is activated by default in Jaws.

**Comment by feerrenrut** [comment]
>@feerrenrut could you please share your thoughts about my concerns in #11906 (comment)

@leonardder I might be misreading your comment, apologies if so! I don't understand the concern. You seem to be saying screen layout is not supported in word etc. We propose to disable screen layout (by default, for new users only). This would mean that all applications (by default) will not use screen layout.

I accept that there may be differences in presentation between Word and browsers with screen layout disabled, but I think this is natural as they are different applications.

**Comment by feerrenrut** [comment]
@leonardder I seem to have misunderstood this in my last comment. Disabling screen layout, activates other behavior in NVDA, breaking up content at control boundaries, as well as breaking up long text. This ability is not available in all applications, it is limited only to applications for which we use a virtual buffer. This means there will be inconsistent behavior between applications by default. I think this is acceptable, different applications have feel different already and have very different presentations and workflows.

**Comment by leonardder** [comment]
> 
> 
> @leonardder I seem to have misunderstood this in my last comment. Disabling screen layout, activates other behavior in NVDA, breaking up content at control boundaries, as well as breaking up long text. This ability is not available in all applications, it is limited only to applications for which we use a virtual buffer. This means there will be inconsistent behavior between applications by default. I think this is acceptable, different applications have feel different already and have very different presentations and workflows.

Ah, that's fine with me as long as you agree it's ok we're introducing inconsistency here. I wonder whether in future we could build something like screen layout off for UIA based browse mode as well

**Comment by feerrenrut** [comment]
> I wonder whether in future we could build something like screen layout off for UIA based browse mode as well

Potentially, it would be a big project. It's certainly something that we will keep in mind.

**Comment by XLTechie** [comment]
@feerrenrut  Is this abandoned / needing someone else to work on it?

I see @OzancanKaratas closed it and deleted the branch, but the conversation prior to that doesn't indicate why.
Best I can tell, it was on track, awaiting mainly a fix involving the system tests.

The original issue is still open; users are still confused by screen lay out by default; and other screen readers (all of them that I have tested) still do not have screen layout on by default, if they have it at all. I would very much like to see this changed for new installs.

**Comment by lukaszgo1** [comment]
While I have no idea why @OzancanKaratas closed this PR there are certain things which needs to be done for any new PR in this area to be accepted IMHO:

Firstly #12067 should be fixed and given this very likely requires changes to the C++ code this is going to be a task on its own.
Secondly any fix in this area should ensure that for people who have explicitly set screen layout to enabled  experience remains unchanged.

**Comment by XLTechie** [comment]
> Firstly #12067 should be fixed and given this very likely requires changes to the C++ code this is going to be a task on its own.

While I agree that should be fixed, I do not believe it should be blocking for 
the issue behind this PR. #12067 is ugly and illogical, yes, but I'm not 
convinced it is such a huge bad thing that it prevents the use of 
non-screen-layout mode, or would even be noticed by most users.

> Secondly any fix in this area should ensure that for people who have explicitly set screen layout to enabled experience remains unchanged.

Obviously, but I believe @OzancanKaratas already had that addressed. At least, 
one of the comments above seems to indicate that.
I haven't reviewed the code changes so far, because I'm not sure what's going on 
with this yet, or if I even can do what needs to be done.

I hoped to first understand why this PR was closed.

**Comment by OzancanKaratas** [comment]
I accidentally deleted the branch I used for this pull request. You can use the patch file of this pull request in another work.

[Patch file](https://github.com/nvaccess/nvda/pull/11906.patch)

**Comment by XLTechie** [comment]
It is your work; I'm sure we would all be happy for you to continue it if you 
are willing.

I was only going to look into it to keep it from being abandoned, but if you 
still can work on it, you are much more up to date on what must be done.

**Comment by oysteinmoseng** [comment]
Is there any movement on this?
